org.jacoco.agent.rt.test/src/org/jacoco/agent/rt/internal/AgentTest.java

```diff
@@ -16,176 +16,239 @@
 	}
 
 	@Test
-	public void testNoSessionId() throws Exception {
-		Agent agent = new Agent(options, this);
+	public void startup_should_register_mbean_when_enabled() throws Exception {
+		options.setJmx(true);
+		Agent agent = createAgent();
 
-		final String defaultId = agent.getData().getSessionId();
+		agent.startup();
+
+		ObjectName objectName = new ObjectName("org.jacoco:type=Runtime");
+		final MBeanServer server = ManagementFactory.getPlatformMBeanServer();
+		assertEquals(JaCoCo.VERSION,
+				server.getAttribute(objectName, "Version"));
 
+		// cleanup as MBean is registered globally
+		agent.shutdown();
+	}
+
+	@Test(expected = InstanceNotFoundException.class)
+	public void startup_should_not_register_mbean_when_disabled()
+			throws Exception {
+		Agent agent = createAgent();
 		agent.startup();
 
-		assertFalse(defaultId.equals(agent.getData().getSessionId()));
-		assertNull(exception);
+		ObjectName objectName = new ObjectName("org.jacoco:type=Runtime");
+		ManagementFactory.getPlatformMBeanServer().getMBeanInfo(objectName);
 	}
 
 	@Test
-	public void testNoDumpOnExit() throws Exception {
-		options.setDumpOnExit(false);
-		Agent agent = new Agent(options, this);
+	public void shutdown_should_write_execution_data_when_enabled()
+			throws Exception {
+		Agent agent = createAgent();
+		agent.startup();
+
+		agent.shutdown();
+
+		assertEquals(Boolean.FALSE, writeExecutionDataReset);
+		assertNull(loggedException);
+	}
 
+	@Test
+	public void shutdown_should_not_write_execution_data_when_disabled()
+			throws Exception {
+		options.setDumpOnExit(false);
+		Agent agent = createAgent();
 		agent.startup();
+
 		agent.shutdown();
 
-		assertEquals(0, execfile.length());
-		assertNull(exception);
+		assertNull(writeExecutionDataReset);
+		assertNull(loggedException);
 	}
 
 	@Test
-	public void testInvalidExecFile() throws Exception {
-		options.setDestfile(folder.getRoot().getAbsolutePath());
-		Agent agent = new Agent(options, this);
+	public void shutdown_should_log_exception() throws Exception {
+		final Exception expected = new Exception();
+		Agent agent = new Agent(options, this) {
+			@Override
+			IAgentOutput createAgentOutput() {
+				return new IAgentOutput() {
+					public void startup(AgentOptions options,
+							RuntimeData data) {
+					}
 
+					public void shutdown() throws Exception {
+						throw expected;
+					}
+
+					public void writeExecutionData(boolean reset) {
+					}
+				};
+			}
+		};
 		agent.startup();
 
-		assertTrue(exception instanceof IOException);
+		agent.shutdown();
+
+		assertSame(expected, loggedException);
+	}
+
+	@Test(expected = InstanceNotFoundException.class)
+	public void shutdown_should_deregister_mbean_when_enabled()
+			throws Exception {
+		options.setJmx(true);
+		Agent agent = createAgent();
+		agent.startup();
+
+		agent.shutdown();
+
+		ObjectName objectName = new ObjectName("org.jacoco:type=Runtime");
+		ManagementFactory.getPlatformMBeanServer().getMBeanInfo(objectName);
```

Comment:
Combine with `startup_should_register_mbean_when_enabled` ?

Reply:
I personally prefer to test one target/aspect at a time in unit tests. In case of test failures you see which aspects are broken.

Reply:
@marchof in case of failure of MBean startup or shutdown, both tests most likely will be failing, and that's why I asked about combining 😉

Reply:
@Godin Sure. Mostlikely. But could also be that registration works but deregistration does not. I simply want to avoid `should_do_this_and_that` unit tests 😉